desktopexporter/internal/store/util.go

```diff
@@ -0,0 +1,243 @@
+	var pairs []string
+	for k, v := range m {
+		var valStr string
+		switch v := v.(type) {
+		case string:
+			valStr = fmt.Sprintf("'%s'::attribute", escapeString(v))
+		case int, int32, int64:
+			valStr = fmt.Sprintf("%d::attribute", v)
+		case float32, float64:
+			valStr = fmt.Sprintf("%f::attribute", v)
+		case bool:
+			if v {
+				valStr = "true::attribute"
+			} else {
+				valStr = "false::attribute"
+			}
+		case []string:
+			elements := make([]string, len(v))
+			for i, s := range v {
+				elements[i] = fmt.Sprintf("'%s'::attribute", escapeString(s))
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []int64:
+			elements := make([]string, len(v))
+			for i, n := range v {
+				elements[i] = fmt.Sprintf("%d::attribute", n)
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []float64:
+			elements := make([]string, len(v))
+			for i, f := range v {
+				elements[i] = fmt.Sprintf("%f::attribute", f)
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []bool:
+			elements := make([]string, len(v))
+			for i, b := range v {
+				if b {
+					elements[i] = "true::attribute"
+				} else {
+					elements[i] = "false::attribute"
+				}
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		default:
+			valStr = fmt.Sprintf("union_value(str := '%v')", v)
+		}
+		pairs = append(pairs, fmt.Sprintf("'%s': %v", escapeString(k), valStr))
+	}
+	return fmt.Sprintf("MAP{%s}", strings.Join(pairs, ", "))
+}
+
+// Helper function to parse events from raw DuckDB [STRUCT(...)] string format
+func parseRawEvents(rawEvents string) []telemetry.EventData {
+	if rawEvents == "" || rawEvents == "[]" {
+		return []telemetry.EventData{}
+	}
+
+	// Remove outer brackets
+	rawEvents = strings.Trim(rawEvents, "[]")
+	if rawEvents == "" {
+		return []telemetry.EventData{}
+	}
+
+	var events []telemetry.EventData
+	// Split on "}, {" to separate individual events
+	rawEventsList := strings.Split(rawEvents, "}, {")
+
+	for _, rawEvent := range rawEventsList {
+		// Clean up the event string
+		rawEvent = strings.Trim(rawEvent, "{}")
+
+		// Split into fields
+		fields := strings.Split(rawEvent, ", ")
+
+		event := telemetry.EventData{
+			Attributes: make(map[string]interface{}),
+		}
+
+		for _, field := range fields {
+			key, value, found := strings.Cut(field, ": ")
+			if !found {
+				continue
+			}
+
+			key = strings.Trim(key, "'")
+			value = strings.Trim(value, "'")
+
+			switch key {
+			case "name":
+				event.Name = value
+			case "timestamp":
+				// Parse timestamp
+				if t, err := time.Parse("2006-01-02 15:04:05.999999999", value); err == nil {
+					event.Timestamp = t
+				}
+			case "attributes":
+				event.Attributes = parseRawAttributes(value)
+			case "droppedAttributesCount":
+				if count, err := strconv.ParseUint(value, 10, 32); err == nil {
+					event.DroppedAttributesCount = uint32(count)
+				}
+			}
+		}
+
+		events = append(events, event)
+	}
```

Comment:
![a capybara with big cartoonish eyes that are in awe of something beautiful. The capybara is holding its face in its paws just astounded at what it's seeing](https://github.com/user-attachments/assets/2d1fe3bd-1da0-41e7-983b-1de0d88fea37)